README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
Coinprofile
===========

This package allows us to calculate the turning points from the Bry and Boschan (1971) methodology. The main goal of Coinprofile is to build the coincident profile according to Martinez et al (2016).

Installation
------------

You can install the released version of Coinprofile from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("Coinprofile")
```

Example
-------

This is a basic example which shows you how to solve a common problem:

``` r
library(Coinprofile)
## basic example code
set.seed(123)
w <- seq(-3, 7, length.out = 100)
x1 <- sin(pi*w)+rnorm(100,0,0.1)
x2 <- sin(pi*w-0.1)+rnorm(100,0,0.1)
coincident_profile(x1, x2, 4, 5, "name.x", "name.y", TRUE, 1991, 2015, 4, 4)
```

<img src="man/figures/README-example-1.png" width="100%" />

    #> $Profile
    #>        p.value lags
    #> p(-5)  0.78125   -5
    #> p(-4)  0.78125   -4
    #> p(-3)  0.78125   -3
    #> p(-2)  0.78125   -2
    #> p(-1)  1.56250   -1
    #> p(0)  75.00000    0
    #> p(1)  10.93750    1
    #> p(2)   1.56250    2
    #> p(3)   0.78125    3
    #> p(4)   0.78125    4
    #> p(5)   0.78125    5
    #> 
    #> $MainLag
    #>   lag_MaxP-value name.x name.y Amount_TP_used
    #> 1              0 name.x name.y              8

    # In this example x leads y three periods
    set.seed(123)
    w <- seq(-3, 7, length.out = 100)
    x <- sin(pi*w)+rnorm(100,0,0.1)
    y <- sin(pi*w-1)+rnorm(100,0,0.1)
    coincident_profile(x, y, 4, 6, "name.x", "name.y", TRUE, 1991, 2015, 4, 4)

<img src="man/figures/README-example-2.png" width="100%" />

    #> $Profile
    #>         p.value lags
    #> p(-6)   0.78125   -6
    #> p(-5)   0.78125   -5
    #> p(-4)   1.56250   -4
    #> p(-3) 100.00000   -3
    #> p(-2)   0.78125   -2
    #> p(-1)   0.78125   -1
    #> p(0)    0.78125    0
    #> p(1)    0.78125    1
    #> p(2)    0.78125    2
    #> p(3)    0.78125    3
    #> p(4)    0.78125    4
    #> p(5)    0.78125    5
    #> p(6)    0.78125    6
    #> 
    #> $MainLag
    #>   lag_MaxP-value name.x name.y Amount_TP_used
    #> 1             -3 name.x name.y              8


NEWS.md
# Coinprofile 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/Dif_MinMax.R
#' @title Closest differences
#'
#' @description Finds closest differences between two sets of turning points
#' @param dato1 A vector of data
#' @param dato2 A vector of data2
#' @param minmax Indicator variable 1 minimum, 2 maximum
#' @return Differences between turning points
#' @import zoo
#' @export
#' @details From Martinez et al. (2016): "When the number of turning points in the two
#'  time series that are being compared is not the same,
#'  we recommend choosing the pairs of turning points, i.e.
#'  peak–peak and trough–trough, in the following way:
#'  take the first peak of the first time series, then
#'  compute the chronological difference with each one
#'  of the peaks of the second time series and choose
#'  the peak in this time series that leaves the smallest
#'  algebraic difference. Now, consider the second peak
#'  of the first time series and compute the differences
#'  with all the peaks of the second one but the peaks
#'  that precedes the one chosen in the previous step.
#'  Take from the second time series that for which the
#'  difference is the smallest. This procedure for pairing
#'  peaks is repeated until either all the peaks in the first
#'  time series have been considered or all the peaks in the
#'  second series have been exhausted."
#'
#' @examples
#' x <- rnorm(100)
#' y <- rnorm(100)
#' dif_TP(TP_BryBoschan(x),TP_BryBoschan(y))
#'
dif_TP <- function(dato1, dato2, minmax = 1){

  dato1 <- data.frame(dato1)
  dato2 <- data.frame(dato2)
  if(nrow(dato1) > 0 & nrow(dato2) > 0){

    x <- data.frame(dato1[which(dato1$Max_Min == minmax), "Time"])
    y <- data.frame(dato2[which(dato2$Max_Min == minmax), "Time"])

    pegas <- NULL
    if(nrow(x) > 0 & nrow(y) > 0){
      for(i in 1:nrow(y)){
        #i=1
        pega <- cbind(x, c(rep(y[i,],nrow(x))) )
        pegas <- rbind(pegas, pega)
      }

      colnames(pegas) <- c("time","time2")
      pegas$difer <- pegas$time - pegas$time2
      pegas$difer2 <- abs(pegas$time - pegas$time2)

      dejas <- pegas

      dejas2 <- NULL
      for(i in unique(dejas$time) ){
        deja <- dejas[which(dejas$time == i),]
        deja1 <- deja[which(deja$difer2 == min(deja$difer2)),]

        if( nrow(deja1) > 1 ){
          deja2 <- deja1[which(deja1$difer == min(deja1$difer)),]
          dejas2 <- rbind(dejas2, deja2)
        }else if( nrow(deja1) == 1 ){
          dejas2 <- rbind(dejas2, deja1 )
        }
      }
      dejas3 <- NULL
      for(i in unique(dejas2$time2) ){
        deja <- dejas2[which(dejas2$time2 == i),]
        deja1 <- deja[which(deja$difer2 == min(deja$difer2)),]
        if( nrow(deja1) > 1 ){
          deja4 <- deja1[which(deja1$difer == min(deja1$difer)),]
          dejas3 <- rbind(dejas3, deja4)
        }else if( nrow(deja1) == 1 ){
          dejas3 <- rbind(dejas3, deja1 )
        }
      }

      #################################################################################
      #################################################################################
      x <- data.frame(dato2[which(dato2$Max_Min == minmax), "Time"])
      y <- data.frame(dato1[which(dato1$Max_Min == minmax), "Time"])

      pegas <- NULL
      for(i in 1:nrow(y)){
        #i=1
        pega <- cbind(x, c(rep(y[i,],nrow(x))) )
        pegas <- rbind(pegas, pega)
      }

      colnames(pegas) <- c("time","time2")
      pegas$difer <- pegas$time - pegas$time2
      pegas$difer2 <- abs(pegas$time - pegas$time2)

      dejas <- pegas

      dejas4 <- NULL
      for(i in unique(dejas$time) ){
        deja <- dejas[which(dejas$time == i),]
        deja1 <- deja[which(deja$difer2 == min(deja$difer2)),]
        if( nrow(deja1) > 1 ){
          deja2 <- deja1[which(deja1$difer == min(deja1$difer)),]
          dejas4 <- rbind(dejas4, deja2)
        }else if( nrow(deja1) == 1 ){
          dejas4 <- rbind(dejas4, deja1 )
        }
      }
      dejas5 <- NULL
      for(i in unique(dejas4$time2) ){
        deja <- dejas4[which(dejas4$time2 == i),]
        deja1 <- deja[which(deja$difer2 == min(deja$difer2)),]
        if( nrow(deja1) > 1 ){
          deja2 <- deja1[which(deja1$difer == min(deja1$difer)),]
          dejas5 <- rbind(dejas5, deja2)
        }else if( nrow(deja1) == 1 ){
          dejas5 <- rbind(dejas5, deja1 )
        }
      }

      colnames(dejas5) <- c(paste(colnames(dejas3),".2",sep=""))
      dejas6 <- merge(dejas3, dejas5, by.x="time", by.y="time2.2")

      if(nrow(dejas6)>0){
        dejas6$depura <- 0
        for(i in 1:nrow(dejas6)){
          if(dejas6[i,3] == -dejas6[i,6]){
            dejas6[i,8] <- 1
          }
        }

        dejas6 <- dejas6[which(dejas6$depura == 1),-c(ncol(dejas6))]
      }else dejas6 <- NULL
    }else dejas6 <- NULL

  }else dejas6 <- NULL

  return(dejas6)
}# End function







R/coincident_profile.R
#' @title Coincident Profile
#'
#' @description Returns the coincident profile developed by Martinez et al. (2016).
#'  The ideal result is finding the maximum p-value for the lag = 0; otherwise
#'  maximum p-value for negative lags suggest leading from x to y, or maximum
#'  p-value for positive lags suggest leading from y to x.
#' @param x Univariate time series
#' @param y Univariate time series
#' @param frequ Frequency of x and y. x and y must have the same frequency
#' @param MLag Maximum lag for the coincident profile
#' @param nvar1 Name of x
#' @param nvar2 Name of y
#' @param print.graf If TRUE returns a panel 2x2 where the superior panel has both
#'  plots of x and y with their turning points (maximums black dots lines and minimums red dots lines).
#'  The inferior left panel has the matplot of x and y standardized, respectively (x-mean(x))/sd(x).
#'  The inferior right panel has the coincident profile where \emph{TP} shows the number
#'  of turning points used.
#' @param iyear The year of the first observation. A single number
#' @param lyear The year of the last observation. A single number
#' @param imonth The amount of months for the first year. A single number
#' @param lmonth The amount of months for the last year. A single number
#' @param tit1 Title for the plot x
#' @param tit2 Title for the plot y
#' @param tit3 Title for the plot x and y
#' @return The coincident profile
#' @import zoo
#' @importFrom graphics abline axis barplot legend
#'  matplot par plot text
#' @export
#' @details The main output contains two objects: the coincident
#'  profile (Profile) and both the lag which has the highest probability and
#'  the number of turning points considered to the calculus of the
#'  coincident profile (MainLag).
#' @author Wilmer O Martinez R
#' @references Martinez, W and Nieto, Fabio H and Poncela, P (2016)
#'  "Choosing a dynamic common factor as a coincident index",
#'  \emph{Statistics and Probability Letters}, (109), 89-98.
#'  \url{http://dx.doi.org/10.1016/j.spl.2015.11.008}.
#' @references Banerji, A., (1999)
#'  "The lead profile and others non-parametrics tools to evaluate survey
#'  series as leading indicators",
#'  \emph{Survey Data for Industry, Research and Economic Policy,
#'   selected papers presented at the 24th CIRET Conference, Willington, New Zealand.}
#' @examples
#' set.seed(123)
#' w <- seq(-3, 7, length.out = 100)
#' x1 <- sin(pi*w)+rnorm(100,0,0.1)
#' x2 <- sin(pi*w-0.1)+rnorm(100,0,0.1)
#' coincident_profile(x1, x2, 4, 5, "name.x", "name.y", TRUE, 1991, 2015, 4, 4)
#'
#' # In this example x leads y three periods
#' set.seed(123)
#' w <- seq(-3, 7, length.out = 100)
#' x <- sin(pi*w)+rnorm(100,0,0.1)
#' y <- sin(pi*w-1)+rnorm(100,0,0.1)
#' coincident_profile(x, y, 4, 6, "name.x", "name.y", TRUE, 1991, 2015, 4, 4)

coincident_profile <- function(x, y, frequ = 12, MLag = 6, nvar1 = "name.x", nvar2 = "name.y",
                               print.graf = FALSE, iyear = 1, lyear = numeric(), imonth = 12,
                               lmonth = numeric(), tit1 = "Title.x", tit2 = "Title.y",
                               tit3 = "Title.x.y"){

  data1_TP <- TP_BryBoschan(x, frequ = frequ)
  data2_TP <- TP_BryBoschan(y, frequ = frequ)

  comp.min <- dif_TP(data1_TP, data2_TP,1)
  comp.max <- dif_TP(data1_TP, data2_TP,2)

  difer <- c(comp.min[,3],comp.max[,3])
  if(!is.null(difer)){
    difer <- difer[abs(difer) < 10]
    #library(coin)
    #library(exactRankTests)
    if( length(difer) >= 3){
      K <- -MLag:MLag
      #if( any(difer != 0) ){
      pvalor <- NULL
      for(k in K){
        p1 <- exactRankTests::perm.test(difer,  paired=TRUE, alternative="two.side",
                        mu=k, exact=T, conf.int=FALSE, conf.level=0.95)
        pvalor <- rbind(pvalor,cbind(p1$p.value*100,k))
      }
      pvalor <- data.frame(pvalor)
      rownames(pvalor) <- paste("p(", K , ")"  , sep = "")
      colnames(pvalor) <- c("p.value","lags")

      lags1 <- data.frame(pvalor[which(pvalor$p.value == max(pvalor$p.value)), c("lags")])
      colnames(lags1) <- "lags"
      if(nrow(lags1) > 1) lags1 <- lags1[which(lags1$lags == max(lags1$lags)), c("lags")]

      period <- data.frame(cbind(lags1, nvar1, nvar2, length(difer) ))
      colnames(period) <- c("lag_MaxP-value", nvar1, nvar2, "Amount_TP_used")

      if(print.graf == TRUE){

        sequ.1 <- c(round(seq(iyear, lyear, by=1),0))
        freq <- c(imonth, frequ, lmonth)

        sequ <- NULL
        for(i.1 in 1:length(sequ.1)){
          #i.1 =1
          if(i.1 == 1){
            sequ.2 <- c(rep(sequ.1[i.1],freq[1]))
          }else if(i.1 < length(sequ.1)){
            sequ.2 <- c(rep(sequ.1[i.1],freq[2]))
          }else if(i.1 == length(sequ.1)){
            sequ.2 <- c(rep(sequ.1[i.1],freq[3]))
          }
          sequ <- c(sequ, sequ.2)
        }

        tit <- paste("Coincident profile",sep="")

        #windows()
        #quartz(width=10,height=6)

        oldpar <- par(las =1,mfrow=c(2,2))
        on.exit(par(oldpar))

        plot(x, type="l",main=tit1, ylab="", xlab="Months", axes=FALSE)
        axis(2)
        axis(1, 1:length(x),sequ)
        maximos <- data1_TP[which(data1_TP$Max_Min == 2 ), "Time"]
        minimos <- data1_TP[which(data1_TP$Max_Min == 1 ), "Time"]

        abline(v = t(maximos),lty=3)
        abline(v = t(minimos),lty=3,col="red")

        plot(y ,type="l",main=tit2, ylab="", xlab="Months", axes = FALSE)
        axis(2)
        axis(1, 1:length(y),sequ)
        maximos <- data2_TP[which(data2_TP$Max_Min == 2 ), "Time"]
        minimos <- data2_TP[which(data2_TP$Max_Min == 1 ), "Time"]
        abline(v = t(maximos),lty=3)
        abline(v = t(minimos),lty=3,col="red")

        y1 <- cbind((x-mean(x))/sd(x), (y-mean(y))/sd(y))
        matplot(x=(1:length(x)), y1, main=tit3,
                xlab="Months", ylab= "", type ="l", col=c(1,2),lwd = 1.5,
                lend=2, lty=c(1,2),axes =FALSE)
        legend("topleft", c(nvar1, nvar2), pch = 20, col=c(1,2), cex=0.8)
        axis(2,-4:4,-4:4)
        axis(1, 1:length(x),sequ)

        Ft1 <- pvalor$p.value
        barplot(Ft1, xlab ="Lag", main=tit, ylab="p-value", ylim = c(0, 100), names.arg=c(K))
        abline(h=10, col = "red",lwd = 2)
        text(2,90,paste("Tp = ",length(difer),sep=""))
      }
    }else warning("The amount of comparisons is less than 4")
  }else stop("There are not enough turning points to compare")
  return(list("Profile" = pvalor, "MainLag" = period))
} #End function

